[이채빈]Week4

caseV12 opened the pull request:
## 요구사항

### 기본

- [x] [기본]이메일 input에서 focus out 할 때, 값이 없을 경우 input에 빨강색 테두리와 아래에 “이메일을 입력해주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일 input에서 focus out 할 때, 이메일 형식에 맞지 않는 값이 있는 경우 input에 빨강색 테두리와 아래에 “올바른 이메일 주소가 아닙니다.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일 input에서 focus out 일 때, input 값이 test@codeit.com 일 경우 input에 빨강색 테두리와 아래에 “이미 사용 중인 이메일입니다.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]비밀번호 input에서 focus out 할 때, 값이 8자 미만으로 있거나 문자열만 있거나 숫자만 있는 경우, input에 빨강색 테두리와 아래에 “비밀번호는 영문, 숫자 조합 8자 이상 입력해 주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]비밀번호 input과 비밀번호 확인 input의 값이 다른 경우, 비밀번호 확인 input에 빨강색 테두리와 아래에 “비밀번호가 일치하지 않아요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]회원가입을 실행할 경우, 문제가 있는 경우 문제가 있는 input에 빨강색 테두리와 에러 메세지가 보이나요?
- [x] [기본]이외의 유효한 회원가입 시도의 경우, “/folder”로 이동하나요?
- [x] [기본]이메일: test@codeit.com, 비밀번호: codeit101 으로 로그인 시도할 경우, “/folder” 페이지로 이동하나요?
- [x] [기본]비밀번호 input에서 focus out 할 때, 값이 없을 경우 아래에 “비밀번호를 입력해주세요.” 에러 메세지가 보이나요?
- [x] [기본]이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입 되나요?
- [x] [기본]이메일, 비밀번호 input에 에러 관련 디자인을 Components 영역의 에러 케이스로 적용했나요?


### 심화

- [x] [심화]눈 모양 아이콘 클릭시 비밀번호의 문자열이 보이기도 하고, 가려지기도 하나요?
- [x] [심화]비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보이나요?
- [x] [심화]로그인, 회원가입 페이지에 공통적으로 사용하는 로직이 있다면, 반복하지 않고 공통된 로직을 모듈로 분리해 사용했나요?

## 주요 변경사항

- 로그인/회원가입 폼 작성 시 유효성 검증 기능 작성
- 스크립트 모듈화

## 스크린샷


## 멘토에게

- 모듈화를 어떤 식으로 해야할 지 감이 안 잡힙니다.
- github pages 배포 오류 해결 후 배포하겠습니다.

caseV12 committed:
feat: signin.js에 input태그 이벤트리스너 추가

caseV12 committed:
feat: signin 페이지 input 태그 pattern, class 추가

caseV12 committed:
style: sign.css에 input 에러 관련 디자인 추가

caseV12 committed:
feat: signup.js에 validPassCheck 이벤트핸들러 추가

caseV12 committed:
feat: "로그인폼 submit 핸들러 추가"

caseV12 committed:
feat: 회원가입폼 submit 핸들러 추가

caseV12 committed:
feat: 비밀번호 숨김 버튼 토글 기능 추가

caseV12 committed:
refactor: 스크립트 파일 모듈화 및 리팩터링

Tanney-102 commented on scripts/modules/handleErrorClass.js:
```diff
@@ -0,0 +1,22 @@
+const CLASSNAME_ERROR = "error";
+const CLASSNAME_ERROR_MSG = "error-msg";
+
+function addErrorClass(target, msg=""){
+    target.classList.add(CLASSNAME_ERROR)
+    if (msg){
+        const errorMSG = document.createElement("p")
+        errorMSG.classList.add(CLASSNAME_ERROR_MSG)
+        errorMSG.textContent = msg
+        target.parentElement.append(errorMSG)
+    }
+}
```
캬 재사용성이 아주 높네요! 굳굳 👍

Tanney-102 commented on scripts/modules/toggleBtn.js:
```diff
@@ -0,0 +1,8 @@
+function togglePasswordHidden(e){
+    const passwordInput = e.target.parentElement.firstElementChild
```
parent의 first child => 이것도 결국 마크업 구조에 의존하는 셀렉입니다.
지양하는 게 좋겠어요

Tanney-102 commented on scripts/modules/validateSignForm.js:
```diff
@@ -0,0 +1,48 @@
+    const emailInput = e.target
+    const email = e.target.value;
+    removeErrorClass(emailInput)
+    if (email === ""){
+        addErrorClass(emailInput, "이메일을 입력해주세요")
+        return
+    }
+    if (!emailInput.validity.valid){
+        addErrorClass(emailInput, "올바른 이메일 주소가 아닙니다.")
+        return
+    }    
+    if (email === "test@codeit.com"){
+        addErrorClass(emailInput, "이미 사용중인 이메일입니다")
+        return
+    }
```
👍

Tanney-102 commented on scripts/modules/validateSignForm.js:
```diff
@@ -0,0 +1,48 @@
+    }
+}
+
+function validatePasswordCheck(e){
```
아주 마이너 하긴한데 함수이름만 봐서는 어떤 이벤트의 핸들러인지 알기 어렵습니다.
타입이라도 있으면 대충 유추할텐데 js를 사용하는 이상 함수에서 어느정도 표현해줘도 좋을 듯해요.
`validatePasswordOnChange` 라든지 `validationPasswordOnInput` 같은 방식으로요!

Tanney-102 commented on scripts/signin-unrefactored.js:
```diff
@@ -0,0 +1,89 @@
+
+function handleSubmitSignIn(e){
+    e.preventDefault()
+    const email = e.target.email.value
+    const password = e.target.password.value
```
오 역시 웹을 좀 해보신 티가 나는군요

Tanney-102 commented on scripts/signin-unrefactored.js:
이 모듈은 정리 예정이신거죠?

Tanney-102 reviewed (commented):
👍

caseV12 committed:
Refactor: togglePasswordHidden 함수의 마크업 구조로부터 독립성을 향상

caseV12 committed:
Refactor: signForm 유효성 검증 로직 개선

커밋 전 매 이벤트마다 불필요하게 노드를 지우고 생성했던 로직을 없앰.
노드를 제거하는 로직은 validate 모듈의 early return 아래로 내림.
renderErrorMSGNode 함수에서는 기존 노드와 에러메시지 확인 후 이에 맞게 작동하도록 함.
이를 통해 불필요한 동작 개선.

함수명이 이벤트와 연결되도록 개선.

caseV12 committed:
Refactor: 모듈화 전 스크립트 삭제
